iml-gui/crate/src/components/lock_indicator.rs

```diff
@@ -0,0 +1,61 @@
+// Copyright (c) 2020 DDN. All rights reserved.
+// Use of this source code is governed by a MIT-style
+// license that can be found in the LICENSE file.
+
+use crate::{
+    components::{font_awesome, popover, tooltip, Placement},
+    generated::css_classes::C,
+};
+use im::HashSet;
+use iml_wire_types::{warp_drive::Locks, LockChange, LockType, ToCompositeId};
+use seed::{prelude::*, *};
+
+pub(crate) fn lock_indicator<T>(x: &dyn ToCompositeId, locks: &Locks) -> Node<T> {
+    if let Some(lcks) = locks.get(&x.composite_id().to_string()) {
+        if !lcks.is_empty() {
+            let (rw, ro): (HashSet<&LockChange>, HashSet<&LockChange>) =
+                lcks.iter().partition(|l| l.lock_type == LockType::Write);
+
+            let mut tooltip = vec![];
+            let mut popup = vec![];
+
+            let rw_c = rw.len();
```

Comment:
I tried a lot to reduce duplicate, and I believe the result is the least evil, because what is happening here is plural forms with hypothetical i18n plus modifying two "global" structs. Even now the plural forms is not trivial (there are two of them :)) and, I think, this has to be done somehow in general.

Most importantly, tearing off some words, like "read" and "write" from the complete phrases is not worthwhile whatsoever. This could be something like "Write locks: 123", etc., but I prefer human-friendly UI :)

Reply:
Even if you were extract the whole strings (which would be the case in i18n), you'd still have some common code here. 

"{} write locks" / "{} read locks"
"1 write lock" / "1 read lock" 
"Write locks" / "Read locks"

are the variant strings, which would need to be pulled from some i18n lookup. If this were abstracted you would be pushing in the lookups instead of the strings based on r/w type. 

That said, I think DRYing this is a nit anyway.

Reply:
I do agree we need to start thinking about i18n in general fwiw.